### DISCUSSION TITLE
8.0.0 🎉

### DISCUSSION THREAD

**Discussion by mfn** [rebing/graphql-laravel#857] [General]
After quite some RC cycles, the next major release finally arrived!

**Note:** this new release contains a LOT of **BREAKING CHANGES**, please make sure to [review the change changelog](https://github.com/rebing/graphql-laravel/blob/master/CHANGELOG.md#2021-11-15-800) and test before you deploy!

High level changes:
- routing has been rewritten/simplified
- processing of GraphQL requests has been unified to better accommodate for optional features like Automatic Persisted Queries (APQ)
- lots-o-breaking changes and clean ups
- Support for Lumen has been removed

Please find all the details below:

## Breaking changes
- Rewrite and simplify how schemas are handled
  - `\Rebing\GraphQL\GraphQL::$schemas` now only holds `Schema`s and not a
    mixture of strings or arrays
  - `\Rebing\GraphQL\GraphQL::schema()` now only accepts a "schema name", but no
    ad hoc `Schema` or "schema configs". To use ad hoc schemas, use
    `\Rebing\GraphQL\GraphQL::buildSchemaFromConfig()` and
    `\Rebing\GraphQL\GraphQL::addSchema()`
  - `\Rebing\GraphQL\GraphQL::queryAndReturnResult()` (and thus also
    `\Rebing\GraphQL\GraphQL::query()`) does not accept ad hoc schemas via
    `$opts['schema']` anymore; it now only can reference a schema via its name.
  - `\Rebing\GraphQL\GraphQL::addSchema()` now only accept `Schema` objects,
    where before it would support ad hoc schemas via array configuration.
    Use `\Rebing\GraphQL\GraphQL::buildSchemaFromConfig()` for that now.
  - `\Rebing\GraphQL\GraphQL::getSchemaConfiguration()` has been removed due to
    the simplifications.
  - `\Rebing\GraphQL\GraphQL::getNormalizedSchemaConfiguration()` does not
    support ad hoc schemas anymore and only accepts the schema name.
  - `\Rebing\GraphQL\GraphQLServiceProvider::bootSchemas()` has been removed due
    to the simplifications.
    
- The following methods now take a `\Illuminate\Contracts\Config\Repository` as
  second argument:
  - `\Rebing\GraphQL\GraphQL::__construct`
  - `\Rebing\GraphQL\GraphQLServiceProvider::applySecurityRules`
- As part of moving the architecture to an execution based middleware approach,
  the following methods have been removed:
  - `\Rebing\GraphQL\GraphQLController::handleAutomaticPersistQueries` has been
    replaced by the `AutomaticPersistedQueriesMiddleware` middleware
  - `\Rebing\GraphQL\GraphQLController::queryContext` has been
    replaced by the `AddAuthUserContextValueMiddleware` middleware\
    If you relied on overriding `queryContext` to inject a custom context, you
    now need to create your own execution middleware and add to your
    configuration
  - `\Rebing\GraphQL\GraphQLController::executeQuery` has become obsolete, no
    direct replacement.

- Routing has been rewritten and simplified [\#757 / mfn](https://github.com/rebing/graphql-laravel/pull/757)
  - All routing related configuration is now within the top level `route`
    configuration key
  - The following configuration options have been removed:
    - `graphql.routes`\
      It's therefore also not possible anymore to register different routes for
      queries and mutations within a schema. Each schema gets only one route
      (except for the default schema, which is registered for the global prefix
      route as well as under its name).\
      If necessary, this can be emulated with different schemas and multi-level
      paths
  - The following configuration options have been moved/renamed:
    - `graphql.prefix` => `graphql.route.prefix`
    - `graphql.controllers` => `graphql.route.controller`\
      Further, providing a controller action for `query` or `mutation` is not
      supported anymore.
    - `graphql.middleware` => `graphql.route.middleware`
    - `graphql.route_group_attributes` => `graphql.route.group_attributes`
  - The actual routes defined have changed:
    - No more separate routes for the HTTP methods
    - 1 route for each schema + 1 route for the group prefix (default schema)
    - If GraphiQL is enabled: 1 route graphiql route for each schema + 1 for the
      graphiql group prefix (default schema)
    - If provided, the `'method'` argument **must** provide the HTTP method
      verbs in uppercase like `POST` or `GET`, `post` or `get` will **not** work.
  - It's now possible to prevent the registering of any routes by making the top
    level `route` an empty array or null
  - `\Rebing\GraphQL\GraphQL::routeNameTransformer` has been removed
  - It's now possible to register schemas with a `-` in their name
  - Routes are now properly cacheable

- Remove the `\Rebing\GraphQL\GraphQLController::$app`  property [\#755 / mfn](https://github.com/rebing/graphql-laravel/pull/755)\
  Injecting the application container early is incompatible when running within
  an application server like laravel/octane, as it's not guaranteed that the
  container received contains all the bindings. If you relied on this property
  when extending the classes, invoke the container directly via
  `Container::getInstance()`.
  
- Remove deprecated `\Rebing\GraphQL\Support\Type::$inputObject` and `\Rebing\GraphQL\Support\Type::$enumObject` properties [\#752 / mfn](https://github.com/rebing/graphql-laravel/pull/752)\
  Instead in your code, extend `\Rebing\GraphQL\Support\InputType` and `\Rebing\GraphQL\Support\EnumType` directly 
  
- Support for Lumen has been removed
  
- Integrate laragraph/utils RequestParser [\#739 / mfn](https://github.com/rebing/graphql-laravel/pull/739)\
  The parsing of GraphQL requests is now more strict:
  - if you send a `GET` request, the GraphQL query has to be in the query parameters
  - if you send a `POST` request, the GraphQL query needs to be in the body\
    Mixing of either isn't possible anymore
  - batched queries will only work with `POST` requests
    This is due to `RequestParser` using `\GraphQL\Server\Helper::parseRequestParams` which includes this check
  Further:  
  - Drop support for configuration the name of the variable for the variables (`params_key`)
  - `GraphQLUploadMiddleware` has been removed (`RequestParser` includes this functionality)
  - Empty GraphQL queries now return a proper validated GraphQL error
      
- In `\Rebing\GraphQL\GraphQL`, renamed remaining instances of `$params` to `$variables`    
  After switching to `RequestParser`, the support for changing the variable name
  what was supposed to `params_key` has gone and thus the name isn't fitting anymore.
  Also, the default value for `$variables` has been changed to `null` to better
  fit the how `OperationParams` works:
  - old: `public function query(string $query, ?array $params = [], array $opts = []): array`
    new: `public function query(string $query, ?array $variables = null, array $opts = []): array`
  - old: `public function queryAndReturnResult(string $query, ?array $params = [], array $opts = []): ExecutionResult`
    new: `public function queryAndReturnResult(string $query, ?array $variables = null, array $opts = []): ExecutionResult`

  - `\Rebing\GraphQL\Support\ResolveInfoFieldsAndArguments` has been removed
  - `$getSelectFields` closure no longer takes a depth parameter

- The `$args` argument, of the `handle` method of the execution middlewares requires `array` as type.  

### Added
- Command to make an execution middleware [\#772 / mfn](https://github.com/rebing/graphql-laravel/pull/772)
- Command to make a schema configuration [\#830 / matsn0w](https://github.com/rebing/graphql-laravel/pull/830)
- The primary execution of the GraphQL request is now piped through middlewares [\#762 / crissi and mfn](https://github.com/rebing/graphql-laravel/pull/762)\
  This allows greater flexibility for enabling/disabling certain functionality
  as well as bringing in new features without having to open up the library.
- Primarily register \Rebing\GraphQL\GraphQL as service and keep `'graphql'` as alias [\#768 / mfn](https://github.com/rebing/graphql-laravel/pull/768)
- Automatic Persisted Queries (APQ) now cache the parsed query [\#740 / mfn](https://github.com/rebing/graphql-laravel/pull/740)\
  This avoids having to re-parse the same queries over and over again.
- Add ability to detect unused GraphQL variables [\#660 / mfn](https://github.com/rebing/graphql-laravel/pull/660)
- Laravel's `ValidationException` is now formatted the same way as a `ValidationError` [\#748 / mfn](https://github.com/rebing/graphql-laravel/pull/748)
- A few missing typehints (mostly array related) [\#849 / mfn](https://github.com/rebing/graphql-laravel/pull/849)

### Changed
- Internally webonyx query plan feature is now used for retrieving information about a query [\#793 / crissi](https://github.com/rebing/graphql-laravel/pull/793))
- Rewrite and simplify how schemas are handled [\#779 / mfn](https://github.com/rebing/graphql-laravel/pull/779)
- Internally stop using the global `config()` function and preferable use the repository or the Facade otherwise [\#774 / mfn](https://github.com/rebing/graphql-laravel/pull/774)
- Don't silence broken schemas when normalizing them for generating routes [\#766 / mfn](https://github.com/rebing/graphql-laravel/pull/766)
- Lazy loading types has been enabled by default [\#758 / mfn](https://github.com/rebing/graphql-laravel/pull/758)
- Make it easier to extend select fields [\#799 / crissi](https://github.com/rebing/graphql-laravel/pull/799)
- The `$args` argument, of the `handle` method of the execution middlewares requires `array` as type [\#843 / sforward](https://github.com/rebing/graphql-laravel/pull/843)
- Embrace thecodingmachine/safe and use thecodingmachine/phpstan-safe-rule to enforce it [\#851 / mfn](https://github.com/rebing/graphql-laravel/pull/851)
- Don't require a return value for the query option of fields [\#856 / sforward](https://github.com/rebing/graphql-laravel/pull/856)

### Fixed
- Fix `TypeNotFound` when an interface defined after another type where it is used [\#828 / kasian-sergeev](https://github.com/rebing/graphql-laravel/pull/828)

### Removed
- The method `\Rebing\GraphQL\GraphQLServiceProvider::provides` was removed [\#769 / mfn](https://github.com/rebing/graphql-laravel/pull/769)\
  It's only relevant for deferred providers which ours however isn't (and can't
  be made into with the current Laravel architecture).

<hr /><em>This discussion was created from the release <a href='https://github.com/rebing/graphql-laravel/releases/tag/8.0.0'>8.0.0 🎉</a>.</em>

**Comment by mfn** [comment]
Note: I did my best to provide a healthy release, but the amount of (also breaking) changes isn't easy to swallow at first!

I did test drive the 8 cycle from day 0 also in production to find edge cases not covered in tests, but despite the necessary BC changes this has looked good so far.

But again, it's possible something has slipped thus please ask any question here or report bugs via issues https://github.com/rebing/graphql-laravel/issues

Thanks 🙏🏼

**Reply by mnogokotin** [reply]
wow, great )
